opencensus/stats/internal/delta_producer.cc

```diff
@@ -0,0 +1,154 @@
+    const uint64_t index = MeasureRegistryImpl::IdToIndex(measurement.id_);
+    ABSL_ASSERT(index < registered_boundaries_.size());
+    switch (MeasureRegistryImpl::IdToType(measurement.id_)) {
+      case MeasureDescriptor::Type::kDouble:
+        it->second[index].Add(measurement.value_double_);
+        break;
+      case MeasureDescriptor::Type::kInt64:
+        it->second[index].Add(measurement.value_int_);
+        break;
+    }
+  }
+}
+
+void Delta::clear() {
+  registered_boundaries_.clear();
+  delta_.clear();
+}
+
+void Delta::SwapAndReset(
+    std::vector<std::vector<BucketBoundaries>>& registered_boundaries,
+    Delta* other) {
+  registered_boundaries_.swap(other->registered_boundaries_);
+  delta_.swap(other->delta_);
+  delta_.clear();
+  registered_boundaries_ = registered_boundaries;
+}
+
+void Delta::Consume() {
+  for (const auto& tagset_stats : delta_) {
+    for (int i = 0; i < registered_boundaries_.size(); ++i) {
+      const MeasureData& data = tagset_stats.second[i];
+      if (data.count() == 0) {
+        // Shortcut if no data has been recorded for this tagset/measure pair.
+        continue;
+      }
+      // TODO: implement (call into StatsManager).
+    }
+  }
+}
+
+DeltaProducer* DeltaProducer::Get() {
+  static DeltaProducer* global_delta_producer = new DeltaProducer();
+  return global_delta_producer;
+}
+
+void DeltaProducer::AddMeasure() {
+  delta_mu_.Lock();
+  absl::MutexLock harvester_lock(&harvester_mu_);
+  registered_boundaries_.push_back({});
+  SwapDeltas();
+  delta_mu_.Unlock();
+  ConsumeLastDelta();
+}
+
+void DeltaProducer::AddBoundaries(uint64_t index,
+                                  const BucketBoundaries& boundaries) {
+  delta_mu_.Lock();
+  auto& measure_boundaries = registered_boundaries_[index];
+  if (std::find(measure_boundaries.begin(), measure_boundaries.end(),
+                boundaries) == measure_boundaries.end()) {
+    absl::MutexLock harvester_lock(&harvester_mu_);
+    measure_boundaries.push_back(boundaries);
+    SwapDeltas();
+    delta_mu_.Unlock();
+    ConsumeLastDelta();
+  } else {
+    delta_mu_.Unlock();
+  }
+}
+
+void DeltaProducer::Record(std::initializer_list<Measurement> measurements,
+                           TagSet tags) {
+  absl::MutexLock l(&delta_mu_);
+  active_delta_.Record(measurements, std::move(tags));
+}
+
+void DeltaProducer::Flush() {
+  delta_mu_.Lock();
+  absl::MutexLock harvester_lock(&harvester_mu_);
+  SwapDeltas();
+  delta_mu_.Unlock();
+  ConsumeLastDelta();
+}
+
+DeltaProducer::DeltaProducer()
+    : harvester_thread_(&DeltaProducer::RunHarvesterLoop, this) {}
+
+void DeltaProducer::SwapDeltas() {
+  active_delta_.SwapAndReset(registered_boundaries_, &last_delta_);
+}
+
+void DeltaProducer::ConsumeLastDelta() { last_delta_.Consume(); }
+
+void DeltaProducer::RunHarvesterLoop() {
+  absl::Time next_harvest_time = absl::Now() + harvest_interval_;
+  while (true) {
+    absl::SleepFor(next_harvest_time - absl::Now());
```

Comment:
Call absl::Now() only once per loop.

Reply:
Done. (A bit complicated because there is a sleep between the two original calls, and we cannot assume that the sleep expires at next_harvest_time because the harvest may have taken longer than the harvest_interval_. But still probably better, given how slow Now() has seemed.)